### PR TITLE
feat: results of rust ethereum clients research

### PR DESCRIPTION
**Pull request by mateuszm-arianelabs**
**Description**:
The result of recent research on the rust ethereum client, it contains three libraries: web3rs, ethers.rs, alloy-core. The initial research of each library is located in an associated folder in the `tools/` directory.

**Related issue(s)**:
[#2772](https://github.com/hashgraph/hedera-json-rpc-relay/issues/2772)

### CODE REVIEW

File: tools/alloy-core-rs-example/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+target
```
Comment (acuarica) [inline tools/alloy-core-rs-example/.gitignore:1]: Should we ignore `.env` as well?

Applies also for the others `.gitignore`.

File: tools/alloy-core-rs-example/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+    let contract_address = receipt.contract_address().expect("Failed to get contract address");
+    let contract = Counter::new(contract_address, &provider);
+    println!("Deployed contract at address: {}", contract.address());
+
+    // Set number
+    let builder = contract.setNumber(U256::from(42));
+    let tx_hash = builder.send().await?.watch().await?;
+
+    println!("Set number to 42: {tx_hash}");
```
Comment (acuarica) [inline tools/alloy-core-rs-example/src/main.rs:68]:
```suggestion
    println!("Set number to 42: TX {tx_hash}");
```

File: tools/alloy-core-rs-example/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    providers::{Provider, ProviderBuilder},
+    rpc::types::TransactionRequest,
+    signers::local::PrivateKeySigner,
+    sol,
+};
+use eyre::Result;
+use std::env;
+use dotenv::dotenv;
+
+sol! {
```
Comment (acuarica) [inline tools/alloy-core-rs-example/src/main.rs:15]: Bytecode is already embedded in `main`, where is this macro being used?

File: tools/ethers-rs-example/src/greeter.json [hunk 4]
```diff
@@ -0,0 +1 @@
...PUSH1 0x1 ADD PUSH2 0x29A JUMP INVALID LOG2 PUSH5 0x6970667358 0x22 SLT KECCAK256 DUP12 SWAP2 PUSH2 0xDFD1 SWAP6 0xD5 CALLDATASIZE XOR SWAP5 0x2A PUSH19 0xA3B481D61A7B142DE919925A0B34F9C986E570 PUSH31 0x64736F6C634300070600330000000000000000000000000000000000000000 ","sourceMap":"70:316:0:-:0;;;123:74;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;-1:-1:-1;123:74:0;;;;;;;;;;-1:-1:-1;123:74:0;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;-1:-1:-1;123:74:0;;-1:-1:-1;;170:20:0;;;;-1:-1:-1;170:8:0;;:20;;;;;:::i;:::-;;123:74;70:316;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;-1:-1:-1;70:316:0;;;-1:-1:-1;70:316:0;:::i;:::-;;;:::o;:::-;;;;;;;;;;;;;;;;;;;;;;"},"abi":[{"inputs":[{"internalType":"string","name":"_greeting","type":"string"}],"stateMutability":"nonpayable","type":"constructor"},{"inputs":[],"name":"greet","outputs":[{"internalType":"string","name":"","type":"string"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"string","name":"_greeting","type":"string"}],"name":"setGreeting","outputs":[],"stateMutability":"nonpayable","type":"function"}]}
```
Comment (acuarica) [inline tools/ethers-rs-example/src/greeter.json:1]: Maybe we should add comment on how to generate this?

File: tools/web3rs-example/src/SimpleStorage.bin [hunk 5]
```diff
@@ -0,0 +1 @@
+6060604052341561000c57fe5b5b60c68061001b6000396000f30060606040526000357c0100000000000000000000000000000000000000000000000000000000900463ffffffff16806360fe47b11460445780636d4ce63c146061575bfe5b3415604b57fe5b605f60048080359060200190919050506084565b005b3415606857fe5b606e608f565b6040518082815260200191505060405180910390f35b806000819055505b50565b600060005490505b905600a165627a7a72305820616d9257b411248095799f7dd90840e3b07ae5c3b6083c0d78d14d826122d3c40029
```
Comment (acuarica) [inline tools/web3rs-example/src/SimpleStorage.bin:1]: Also some comment on how this was generated.

**Review by acuarica — CHANGES_REQUESTED** [review]
lg

File: tools/ethers-rs-example/src/greeter.json (hunk 4 above)
Comment (mateuszm-arianelabs) [reply tools/ethers-rs-example/src/greeter.json]: I used one contract for all clients and wrote in the README how to generate these artifacts.

File: tools/alloy-core-rs-example/src/main.rs (hunk 3 above)
Comment (acuarica) [reply tools/alloy-core-rs-example/src/main.rs]: If this is not used, we should remove it.

**Review by acuarica — CHANGES_REQUESTED** [review]

File: tools/alloy-core-rs-example/src/main.rs (hunk 3 above)
Comment (arianejasuwienas) [reply tools/alloy-core-rs-example/src/main.rs]: @acuarica 
This sol! macro generates a struct that corresponds to the Solidity contract defined in the JSON artifact. We are using it on this [line](https://github.com/hashgraph/hedera-json-rpc-relay/pull/2773/files#diff-b78e3f872da383914a721d1256e7222ffe1180b3ef95af16f9024d6c58519f5eR44) . I believe that in the current version of this PR, the bytecode is no longer embedded in the main.rs file. However, even if it were, this struct could still be used to interact with the corresponding Smart Contract.

What concerns me more is that I don't see any actual usage of either ReceiptResponse or TransactionBuilder. We should remove them from the imports, @mateuszm-arianelabs .

Comment (acuarica) [reply tools/alloy-core-rs-example/src/main.rs]: Make sense, thanks @arianejasuwienas for the clarification.

Comment (mateuszm-arianelabs) [reply tools/alloy-core-rs-example/src/main.rs]: Yes, we need this code to correspond to the Solidity contract. I have removed unused imports and reformatted the code.

**Review by acuarica — APPROVED** [review]
lg

**Review by ebadiere — APPROVED** [review]
LG

**Review by jeromy-cannon — APPROVED** [review]

**Comment by arianejasuwienas** [comment]
Moved to: https://github.com/hashgraph/hedera-json-rpc-relay/pull/2999

**Pull request closed without merging**